### PR TITLE
make TestListPager_EachListItem rework

### PR DESCRIPTION
**Pull request by Abirdcfly**
Signed-off-by: Abirdcfly <fp544037857@gmail.com>

#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
`TestListPager_EachListItem` not work fine.
#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
None
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by Abirdcfly** [comment]
/hold
I need some advice to determine whether I **should fix the tests or the source code**.

Here are 2 tests where the results look different than expected:
1. When calling cancel, should it return exactly the number at the time of cancel (3), or the number of chunks (10)?
2. The other issue is the number returned when call panic.

@aojea https://github.com/kubernetes/kubernetes/pull/111228#discussion_r923865802
@caesarxuchao https://github.com/kubernetes/kubernetes/pull/111241#issuecomment-1188706400
@jpbetz https://github.com/kubernetes/kubernetes/pull/75849

**Comment by aojea** [comment]
not familiar with this code , maybe @wojtek-t or @liggitt ?

**Comment by Abirdcfly** [comment]
/cc @liggitt

The last cc didn't seem to work

**Comment by liggitt** [comment]
I agree the Fatal-or-return logic was wrong and was masking issues with the test

I'll defer to @jpbetz' review of the test changes themselves

File: staging/src/k8s.io/client-go/tools/pager/pager_test.go [hunk 1]
```diff
@@ -301,14 +301,14 @@ func TestListPager_EachListItem(t *testing.T) {
 		{
 			name:                "cancel context while processing",
 			fields:              fields{PageSize: 10, PageFn: (&testPager{t: t, expectPage: 10, remaining: 51, rv: "rv:20"}).PagedList},
-			want:                list(3, "rv:20"), // all the items <= the one the processor returned an error on should have been visited
+			want:                list(10, "rv:20"), // all the items <= the one the processor returned an error on should have been visited
```
Comment (jpbetz) [inline staging/src/k8s.io/client-go/tools/pager/pager_test.go:304]: This looks right. The whole `PageSize` worth of items got returned.

File: staging/src/k8s.io/client-go/tools/pager/pager_test.go [hunk 2]
```diff
@@ -301,14 +301,14 @@ func TestListPager_EachListItem(t *testing.T) {
...
 			cancelContextOnItem: 3,
 		},
 		{
 			name:      "panic processing item",
 			fields:    fields{PageSize: 10, PageFn: (&testPager{t: t, expectPage: 10, remaining: 51, rv: "rv:20"}).PagedList},
-			want:      list(3, "rv:20"), // all the items <= the one the processor returned an error on should have been visited
+			want:      list(51, "rv:20"), // all the items <= the one the processor returned an error on should have been visited
 			wantPanic: true,
 		},
 	}
```
Comment (jpbetz) [inline staging/src/k8s.io/client-go/tools/pager/pager_test.go:314]: This test is just plain wrong. It should have had `processorPanicOnItem: 3` set. It overlaps with the above `error processing item` and so should just be removed. My apologies.

File: staging/src/k8s.io/client-go/tools/pager/pager_test.go [hunk 3]
```diff
@@ -345,8 +345,7 @@ func TestListPager_EachListItem(t *testing.T) {
 				err = p.EachListItem(ctx, metav1.ListOptions{}, fn)
 			}()
 			if (panic != nil) && !tt.wantPanic {
-				t.Fatalf(".EachListItem() panic = %v, wantPanic %v", panic, tt.wantPanic)
-			} else {
+				t.Errorf(".EachListItem() panic = %v, wantPanic %v", panic, tt.wantPanic)
```
Comment (jpbetz) [inline staging/src/k8s.io/client-go/tools/pager/pager_test.go:348]: Thanks for catching this.

**Comment by Abirdcfly** [comment]
/unhold

ready for review.

**Comment by jpbetz** [comment]
/lgtm

Needs approver.

**Comment by aojea** [comment]
> /lgtm
> 
> Needs approver.

/Assign @lavalamp

**Comment by Abirdcfly** [comment]
kindly ping @lavalamp @caesarxuchao for approve, because master is open again for 1.26 =.=

**Comment by aojea** [comment]
/assign @liggitt 

for approval, 

> /lgtm
> 
> Needs approver.

**Pull request merged**